soh/soh/Enhancements/custom_message/CustomMessageManager.cpp

```diff
@@ -0,0 +1,150 @@
+                                       { "Ô", 0x8A }, { "Ö", 0x8B }, { "Ù", 0x8C }, { "Û", 0x8D }, { "Ü", 0x8E },
+                                       { "ß", 0x8F }, { "à", 0x90 }, { "á", 0x91 }, { "â", 0x92 }, { "ä", 0x93 },
+                                       { "ç", 0x94 }, { "è", 0x95 }, { "é", 0x96 }, { "ê", 0x97 }, { "ë", 0x98 },
+                                       { "ï", 0x99 }, { "ô", 0x9A }, { "ö", 0x9B }, { "ù", 0x9C }, { "û", 0x9D },
+                                       { "ü", 0x9E } };
+    this->colors = { { "w", QM_WHITE }, { "r", QM_RED },  { "g", QM_GREEN },  { "b", QM_BLUE },
+                     { "c", QM_LBLUE }, { "p", QM_PINK }, { "y", QM_YELLOW }, { "B", QM_BLACK } };
+}
+
+CustomMessageManager::~CustomMessageManager() {
+    this->textBoxSpecialCharacters.clear();
+    this->colors.clear();
+    this->messageTables.clear();
+}
+
+void CustomMessageManager::ReplaceSpecialCharacters(std::string& string) {
+    // add special characters
+    for (auto specialCharacterPair : this->textBoxSpecialCharacters) {
+        size_t start_pos = 0;
+        std::string textBoxSpecialCharacterString = ""s;
+        textBoxSpecialCharacterString += specialCharacterPair.second;
+        while ((start_pos = string.find(specialCharacterPair.first, 0)) != std::string::npos) {
+            string.replace(start_pos, specialCharacterPair.first.length(), textBoxSpecialCharacterString);
+            start_pos += textBoxSpecialCharacterString.length();
+        }
+    }
+}
+
+void CustomMessageManager::ReplaceColors(std::string& string) {
+    for (auto colorPair : colors) {
+        std::string textToReplace = "%";
+        textToReplace += colorPair.first;
+        size_t start_pos = 0;
+        while ((start_pos = string.find(textToReplace)) != std::string::npos) {
+            string.replace(start_pos, textToReplace.length(), COLOR(colorPair.second));
+            start_pos += textToReplace.length();
+        }
+    }
+}
+
+void CustomMessageManager::FormatCustomMessage(std::string& message, ItemID iid) {
+    message.insert(0, ITEM_OBTAINED(iid));
+    size_t start_pos = 0;
+    std::replace(message.begin(), message.end(), '&', NEWLINE()[0]);
+    while ((start_pos = message.find('^', start_pos)) != std::string::npos) {
+        message.replace(start_pos, 1, WAIT_FOR_INPUT() + ITEM_OBTAINED(iid));
+        start_pos += 3;
+    }
+    std::replace(message.begin(), message.end(), '@', PLAYER_NAME()[0]);
+    ReplaceSpecialCharacters(message);
+    ReplaceColors(message);
+    message += MESSAGE_END();
+}
+
+void CustomMessageManager::FormatCustomMessage(std::string& message) {
+    size_t start_pos = 0;
+    std::replace(message.begin(), message.end(), '&', NEWLINE()[0]);
+    std::replace(message.begin(), message.end(), '^', WAIT_FOR_INPUT()[0]);
+    std::replace(message.begin(), message.end(), '@', PLAYER_NAME()[0]);
+    ReplaceSpecialCharacters(message);
+    ReplaceColors(message);
+    message += MESSAGE_END();
+}
```

Comment:
What default value would I set ItemID to? I don't think it can be null, and I'm pretty sure 0 is a valid ItemID (Deku Sticks I think?) -1? This might also have to be refactored again later when the GetItemEntries table gets refactored/enhanced.

Reply:
maybe `-1` works? if so then we could have something like
```cpp
void CustomMessageManager::FormatCustomMessage(std::string& message) {
    FormatCustomMessage(message, -1);
}
```

but i'm not sure if it'd end up being cleaner or not, as i said before, bouncing back and forth

Reply:
I was thinking something more like:
```c++
void CustomMessageManager::FormatCustomMessage(std::string& message, ItemID iid = -1) {
    if (iid != -1) {
        message.insert(0, ITEM_OBTAINED(iid));
    }
    size_t start_pos = 0;
    std::replace(message.begin(), message.end(), '&', NEWLINE()[0]);
    if (iid != -1) {
        while ((start_pos = message.find('^', start_pos)) != std::string::npos) {
            message.replace(start_pos, 1, WAIT_FOR_INPUT() + ITEM_OBTAINED(iid));
            start_pos += 3;
        }
    } else {
        std::replace(message.begin(), message.end(), '^', WAIT_FOR_INPUT()[0]);
    }
    std::replace(message.begin(), message.end(), '@', PLAYER_NAME()[0]);
    ReplaceSpecialCharacters(message);
    ReplaceColors(message);
    message += MESSAGE_END();
}
```
Not sure if that syntax is exactly right for optionally specifiable arguments with default values in C++, but something like that. `ItemID` also might need to be `uint8_t` or something for that to work, which might be better anyway if we want to use it for other icons.

Alternatively this might even need to go away altogether in favor of some mod-developer friendly way of specifying the icon to show within the strings themselves (which technically they can already do if they know the control codes, that's what the altar text is doing currently).

Reply:
actually yeah, replacing the itemid stuff with a more user/modder friendly way to do icons in the strings is definitely the move

Reply:
Any suggestions on the syntax? I'm thinking something along the lines of `${iconID}`, but maybe something else would be better?